xpu: implement xpu serialization

dvrogozh opened the pull request:
Fixes: #125529

BC-breaking note:
The deprecated "async" argument to the Storage.cuda and Storage.hpu has been removed. Use non_blocking instead.

CC: @jbschlosser, @frank-wei @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @albanD


cc @frank-wei @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @EikanWang @fengyuan14 @guangyey

dvrogozh commented:
@jbschlosser - can you, please, also help to triage this PR and add appropriate reviewers and labels?

guangyey commented on torch/serialization.py:
```diff
@@ -269,6 +270,13 @@ def _meta_tag(obj):
     if obj.device.type == 'meta':
         return 'meta'
 
+def _meta_tag(obj):
```
typo?

dvrogozh replied:
yep. fixed.

guangyey commented on torch/serialization.py:
```diff
@@ -353,6 +361,35 @@ def _meta_deserialize(obj, location):
+def _xpu_deserialize(obj, location):
+    if location.startswith('xpu'):
+        xpu = getattr(torch, "xpu", None)
+        assert xpu is not None, "XPU device module is not loaded"
```
`_cuda_deserialize`, `hpu_deserialize`, and `_xpu_deserialize` have the same logic. Please address @albanD  [comment](https://github.com/pytorch/pytorch/issues/125529#issuecomment-2096840784) to make the code more generic.

dvrogozh replied:
Just want to note that `_privateuse1_deserialize` is another one with the same logic. Generalization will need to cover it as well.

dvrogozh replied:
Ok, I generalized `_privateuse1_deserialize` to cover cuda, hpu, xpu and privateuse1 backends.

guangyey commented on torch/_utils.py:
```diff
@@ -120,6 +120,36 @@ def _cuda(self, device=None, non_blocking=False, **kwargs):
             untyped_storage.copy_(self, non_blocking)
             return untyped_storage
 
+def _xpu(self, device=None, non_blocking=False, **kwargs):
```
`_xpu`, `_hpu` and `_cuda` have the same logic.

dvrogozh replied:
@guangyey, @EikanWang, @albanD : thank you for your guidance. I caught an idea and work on implementation. Hopefully will update PR tomorrow.

dvrogozh committed:
torch: generalize serialization path

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh committed:
xpu: implement xpu serialization

Fixes: #125529

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

dvrogozh commented on torch/jit/_serialization.py:
```diff
@@ -179,7 +179,7 @@ def validate_map_location(map_location=None):
 
     if str(map_location).startswith("cuda"):
-        validate_cuda_device(map_location)
+        validate_device(map_location)
```
Should this call be applied to other backends as well? If yes, should this be limited to `['cuda', 'hpu', 'xpu']` (I am not sure that this function will work properly for other backends).

dvrogozh replied:
I exposed `validate_cuda_device` for backward compatibilty as a tiny wrapper on top of actual `_validate_device`. This change got reverted to original version. Question still remains however whether this call should be generalized to other backends.

dvrogozh commented:
Ok, I update PR and tried to generalize implementations. I hope that's closer now to what you anticipate. Please, take a look and provide further suggestions.

guangyey commented on torch/serialization.py:
```diff
@@ -397,26 +339,26 @@ def _validate_privateuse1_device(location, backend_name):
-        device_index = _validate_privateuse1_device(location, backend_name)
-        return getattr(obj, backend_name)(device_index)
+def _deserialize(obj, location):
+    for backend_name in ['cuda', 'hpu', 'xpu', torch._C._get_privateuse1_backend_name()]:
```
We should not limit device type for `_deserialize`. It is better to accept a parameter `device_type` in `_deserialize` to make the code more generic and can be shared across other backends. So suggest to change to `deserialize(device, obj, location)`

dvrogozh replied:
done

guangyey commented on torch/serialization.py:
```diff
@@ -397,26 +339,26 @@ def _validate_privateuse1_device(location, backend_name):
 
 register_package(10, _cpu_tag, _cpu_deserialize)
-register_package(20, _cuda_tag, _cuda_deserialize)
+register_package(20, _cuda_tag, _deserialize)
```
```suggestion
register_package(20, _cuda_tag, functools.partial(_deserialize, "cuda"))
```

dvrogozh replied:
done

guangyey commented on torch/_utils.py:
```diff
@@ -52,71 +52,40 @@ def _type(self, dtype=None, non_blocking=False, **kwargs):
     no copy is performed and the original object is returned.
 
     Args:
-        device (int): The destination HPU id. Defaults to the current device.
```
update the doc

dvrogozh replied:
I changed the doc. You probable accidentally looked into prev. version.

guangyey commented on torch/storage.py:
```diff
@@ -826,6 +827,13 @@ def hpu(self, device=None, non_blocking=False, **kwargs) -> T:  # type: ignore[m
+        _warn_typed_storage_removal()
+        if self.dtype in [torch.quint8, torch.quint4x2, torch.quint2x4, torch.qint32, torch.qint8]:
+            raise RuntimeError("Cannot create XPU storage with quantized dtype")
+        xpu_storage: torch.UntypedStorage = self._untyped_storage.xpu(device, non_blocking, **kwargs)
```
for MYPY issue `"UntypedStorage" has no attribute "xpu" `, please refer to ["UntypedStorage" has no attribute "xpu" ](https://github.com/pytorch/pytorch/blob/88fbe79550b41e94d4b78bb5be3afff10cb0ef99/torch/storage.py#L41)

dvrogozh replied:
I fixed it locally. Will push the update as soon as we will close on bc-linter story.

dvrogozh replied:
pushed a fix.

dvrogozh commented on torch/serialization.py:
```diff
@@ -46,8 +47,7 @@
     'check_module_version_greater_or_equal',
-    'validate_cuda_device',
-    'validate_hpu_device',
+    'validate_device',
```
I added backward compatibility wrappers and pushed the change. If you believe this should be done in some other way, let me know.

dvrogozh commented:
Just fixed new linter issues which previously were masked. No functional changes, only `lintrunner -a`.

guangyey commented on test/test_serialization.py:
```diff
@@ -514,24 +514,33 @@ def load_bytes():
         gpu_last_map_locations = [
             f'cuda:{torch.cuda.device_count() - 1}',
         ]
+        xpu_0_map_locations = [
```
Maybe we can unify `gpu_0_map_locations` and `xpu_0_map_locations` to a `def generate_map_location(device_type):` to generate the list.

dvrogozh replied:
done

guangyey commented on test/test_serialization.py:
```diff
@@ -514,24 +514,33 @@ def load_bytes():
-        check_map_locations(cpu_map_locations, torch.FloatTensor, torch.device('cpu'))
+        check_map_locations(cpu_map_locations, torch.float, torch.device('cpu'))
         if torch.cuda.is_available():
-            check_map_locations(gpu_0_map_locations, torch.cuda.FloatTensor, torch.device('cuda', 0))
```
nice~

guangyey commented on test/test_serialization.py:
```diff
@@ -514,24 +514,33 @@ def load_bytes():
+                torch.float,
                 torch.device('cuda', torch.cuda.device_count() - 1)
             )
+        if torch.xpu.is_available():
```
may I know why is it not required of `gpu_last_map_locations` for xpu?

dvrogozh replied:
added

guangyey commented on torch/serialization.py:
```diff
@@ -269,6 +270,9 @@ def _meta_tag(obj):
     if obj.device.type == 'meta':
         return 'meta'
 
+def _xpu_tag(obj):
```
Can we unify all `_xxx_tag` to
```python
def _backend_tag(backend, obj):
    if obj.device.type == backend:
        if obj.device.index is None:
            return backend
        else:
            return backend + ':' + str(obj.device.index)
```
```

dvrogozh replied:
cpu, mps and meta doesn't currently check `obj.device.index`. Does they have it always `=0`?

dvrogozh replied:
Generalized cuda, hpu, xpu and privateuse1. I left cpu, mps and meta unchanged - see question above.

guangyey replied:
Yes, you are right~

albanD commented on test/test_xpu.py:
```diff
@@ -236,6 +237,39 @@ def convert_boolean_tensors(x):
 
+    def test_serialization_array_with_storage(self):
+        x = torch.randn(5, 5).xpu()
+        y = torch.IntTensor(2, 5).fill_(0).xpu()
```
```suggestion
        y = torch.zeros(2, 5, dtype=torch.int, device="xpu")
```

dvrogozh replied:
works, will apply.

albanD commented on test/test_xpu.py:
```diff
@@ -236,6 +237,39 @@ def convert_boolean_tensors(x):
+    def test_serialization_array_with_storage(self):
+        x = torch.randn(5, 5).xpu()
+        y = torch.IntTensor(2, 5).fill_(0).xpu()
+        q = [x, y, x, y.storage()]
```
```suggestion
        q = [x, y, x, y.untyped_storage()]
```
This is going to raise a warning otherwsie

dvrogozh replied:
If applied this change causes runtime error:
```
Traceback (most recent call last):
  File "/home/dvrogozh/git/pytorch/torch/testing/_internal/common_utils.py", line 2763, in wrapper
    method(*args, **kwargs)
  File "/home/dvrogozh/git/pytorch/torch/testing/_internal/common_device_type.py", line 432, in instantiated_test
    raise rte
  File "/home/dvrogozh/git/pytorch/torch/testing/_internal/common_device_type.py", line 419, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/dvrogozh/git/pytorch/test/test_xpu.py", line 245, in test_serialization_array_with_storage
    torch.save(q, f)
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 613, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol, _disable_byteorder_record)
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 825, in _save
    pickler.dump(obj)
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 803, in persistent_id
    raise RuntimeError(
RuntimeError: Cannot save multiple tensors or storages that view the same data as different types
```
Change was:
```
diff --git a/test/test_xpu.py b/test/test_xpu.py
index 2526b65e589..85212d919f9 100644
--- a/test/test_xpu.py
+++ b/test/test_xpu.py
@@ -239,8 +239,8 @@ if __name__ == "__main__":

     def test_serialization_array_with_storage(self):
         x = torch.randn(5, 5).xpu()
-        y = torch.IntTensor(2, 5).fill_(0).xpu()
-        q = [x, y, x, y.storage()]
+        y = torch.zeros(2, 5, dtype=torch.int, device="xpu")
+        q = [x, y, x, y.untyped_storage()]
         with tempfile.NamedTemporaryFile() as f:
             torch.save(q, f)
             f.seek(0)
```

dvrogozh replied:
It seems this will need to be addressed after/during migration of tensor underlying storages to `untyped_storage()`. I left his unaddressed. If you have idea how to address this now - let me know.

albanD replied:
Ok, let's punt this to a follow up then!

albanD commented on test/test_xpu.py:
```diff
@@ -236,6 +237,39 @@ def convert_boolean_tensors(x):
+        self.assertTrue(isinstance(q_copy[3], torch.storage.TypedStorage))
+        self.assertTrue(isinstance(q_copy[3]._untyped_storage, torch.UntypedStorage))
+        q_copy[1].fill_(10)
+        self.assertEqual(q_copy[3], torch.IntStorage(10).fill_(10).xpu())
```
It would be best not to use the dtype-specific classes like FloatTensor / FloatStorage.
For this one, I would create a new Tensor and get the storage from it.

dvrogozh replied:
We already have such a tensor (`y`), so I just reused it.

albanD commented on torch/_utils.py:
```diff
@@ -52,70 +52,49 @@ def _type(self, dtype=None, non_blocking=False, **kwargs):
 
     Args:
-        device (int): The destination HPU id. Defaults to the current device.
+        device (int): The destination device id. Defaults to the current device.
```
I know this is existing code, but I think we should rename this to `device_idx` to make it clear that this is just the device index and not the full device including devicetype

dvrogozh replied:
changed

albanD commented on torch/serialization.py:
```diff
@@ -354,13 +288,13 @@ def _meta_deserialize(obj, location):
-    Register a device_module of privateuse1 by torch._register_device_module.
-    Implement the following methods in device_module like cuda:
-    device_module._utils._get_device_index(location, True),
+    In case of privateuse1 backend, first register a device_module of
+    privateuse1 by torch._register_device_module. Implement the following
```
```suggestion
    In case of privateuse1 backend, your must first register a device_module for
    privateuse1 using torch._register_device_module. Implement the following
```

albanD commented on torch/storage.py:
```diff
@@ -40,6 +40,7 @@ def size(self) -> int:
     def type(self, dtype: _Optional[str] = None, non_blocking: bool = False) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
     def cuda(self, device=None, non_blocking=False, **kwargs) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
     def hpu(self, device=None, non_blocking=False, **kwargs) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
+    def xpu(self, device=None, non_blocking=False, **kwargs) -> T: ...  # type: ignore[empty-body, misc, type-var] # noqa: E704
```
Ho why do we need these new methods? That's unfortunate :/

dvrogozh replied:
> I would still like to discuss why we're adding new xpu() methods on the Storage object? We have all the generic equivalent there to satisfy any code update we might need to do right?

I am following the design architecture pattern which I observe here - add `xpu()` alongside with `cuda()` and `hpu()`. Note that at the moment you don't seem to have _the generic equivalent_ of these methods exposed on the uAPI level. So, with the current design I need to add `xpu()`. Alternate approach might be to expose actual generic uAPI, for example:
```
def backend(self, backend, device_idx=None, non_blocking=False, **kwargs) -> T: ... # type: ignore[empty-body, misc, type-var] # noqa: E704
```
and deprecate `cuda()` and `hpu()`. Eventually, we can add trivial wrappers on top of `backend` for backward compatibility.

Is above what you think about or you have deeper architecture refactoring in mind?

albanD replied:
> Note that at the moment you don't seem to have the generic equivalent of these methods exposed on the uAPI level.

Ho I missed that, that is indeed problematic here.

> Is above what you think about or you have deeper architecture refactoring in mind?

I would like to understand the difference between the backend() method you have and the .to() method we have on Tensor. They look very much the same at a high level.
If they are actually similar, I would go along the idea of https://github.com/pytorch/pytorch/blob/31372fa84293f5bdd0fb06acceb098a5c8187a07/torch/storage.py#L148-L154
But instead passing *args and **kwargs along to the call to to().
This will give us a generic API that is consistent with the Tensor API and that we don't need custom logic to maintain.

I think the 3 options we have here are:
1. Do this refactor to expose a full .to() API in this PR, use that and remove the xpu-specific method
2. Do the refactor in a follow up PR. Still remove the xpu-specific methods from this PR since we know we won't need them
3. Do the refactor in a follow up PR. Keep the xpu-specific APIs here knowing that we will deprecate and remove them once we add the generic API.

I personally think that 2 is the best option. If you do need a shorter-term solution/you won't be able to work on the refactor anytime soon, I could get convinced to do 3.

dvrogozh replied:
We can't remove the xpu-specific methods from this PR, specifically - this `_StorageBase.xpu` method we are discussing here because it's essential to make a `load(map_location=xpu)` to work. Here is how it errors out on this PR if I comment xpu change in torch/storage.py:
```
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1014, in load
    return _load(opened_zipfile,
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1435, in _load
    result = unpickler.load()
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1405, in persistent_load
    typed_storage = load_tensor(dtype, nbytes, key, _maybe_decode_ascii(location))
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1379, in load_tensor
    wrap_storage=restore_location(storage, location),
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 1302, in restore_location
    return default_restore_location(storage, map_location)
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 375, in default_restore_location
    result = fn(storage, location)
  File "/home/dvrogozh/git/pytorch/torch/serialization.py", line 347, in _deserialize
    raise RuntimeError(f'Attempting to load the storages to the {backend_name.upper()} device '
RuntimeError: Attempting to load the storages to the XPU device but torch.storage._StorageBase.xpu() or torch.storage.TypedStorage.xpu() is not generated. Please use torch.utils.generate_methods_for_privateuse1_backend to generate storage.xpu() method first.
```

So, I think that unfortunately option 2 is not possible and we need to decide on option 1 or option 3. That said I agree with your point that `_backend()` method I invented resembles `to()` and if expose uAPI, then it should be named `to()` and considered aligned with existing tensor's `to()`. While `_backend` is internal api, it's name is not that critical, but if exposed it should be `to()`.

I personally don't like to invent an API which will be deprecated next day. So, unless there will be a huge rock under the water, I suggest to look into option 1. Let's see how it will go for me today.

Let me know if I overlooked something in above estimation or exaggerated the problem.

dvrogozh replied:
@albanD : ok, so I've just pushed one more commit which implements `.to()` for storage. Please, take a look if it goes in the right direction.

albanD reviewed (commented):
Thanks a lot for the cleanup!
Small comments only.

FYI @kulinseth it would be good to migrate mps to use this generic code as well when we have a minute.

guangyey commented on test/test_serialization.py:
```diff
@@ -493,6 +493,15 @@ def test_serialization_map_location(self):
 
+        def generate_map_locations(device_type):
+            return [
+                {'cuda:0': device_type + ':0'},
```
```suggestion
                {device_type + ':0': device_type + ':0'},
```

dvrogozh replied:
`cuda:0` on a left side of `{'cuda:0' : device_type + ':0'}` isn't a mistake. The example tensor we read on line 491 above is cuda generated, that's exactly where map_location comes handy - it allows to load tensor to other device memory bypassing original device which might not exist on the moment of execution. As far as I understand `{'cuda:0' : device_type + ':0'}` is a full signature of requested operation - we want tensor which is originally create for cuda to be loaded to other backend.

guangyey replied:
you are right~

guangyey commented:
some nits, otherwise LGTM.

albanD reviewed (commented):
The change to the serialization code looks good.
I would still like to discuss why we're adding new xpu() methods on the Storage object? We have all the generic equivalent there to satisfy any code update we might need to do right?

dvrogozh commented:
@albanD : anything I should address on the last iterations with `.to()` added for storage?

dvrogozh committed:
torch: drop kwargs in storage cuda(), hpu() and to() methods

The might have been long deprecated 'async' parameters passed
in kwargs. Per discussion in code review we've decided it's time
to make a next step and drop it completely.

Signed-off-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

albanD commented on torch/storage.py:
```diff
@@ -826,6 +859,13 @@ def hpu(self, device=None, non_blocking=False, **kwargs) -> T:  # type: ignore[m
 
+    def to(self, device: torch.device, non_blocking: bool = False, **kwargs) -> T:  # type: ignore[type-var]
+        _warn_typed_storage_removal()
+        if self.dtype in [torch.quint8, torch.quint4x2, torch.quint2x4, torch.qint32, torch.qint8]:
+            raise RuntimeError("Cannot create XPU storage with quantized dtype")
```
This needs removing I guess?
The c++ code should throw that error anyways right?

dvrogozh replied:
I think it should, but whether underlying C++ cross will actually do that across 3 backends (cuda, hpu, xpu), I've no idea. Also, this might have been optimization to throw error earlier. I tend to leave this as is unless you will insist. (need to fix "XPU" in the message)

albanD replied:
Was this an existing error message? I can't find any reference to it.
And yes I do expect backends like cuda to definitely implement this properly.

dvrogozh replied:
Actually, since that's a new API (`.to`), I think we can avoid having this thrown relying on c++ code. And if there will be a case when c++ does not throw, this can be considered an error. So, I will remove.

dvrogozh replied:
> Was this an existing error message? I can't find any reference to it.

Sorry, I somehow missed this comment. Yes, it was existing message in `cuda()` and `hpu()` which we generalize into this new `to()`. See:
https://github.com/pytorch/pytorch/blob/3d20a5dd87e9c45e20dbc89210cfc63e7824b2b9/torch/storage.py#L850-L851

albanD replied:
Ho ok!
Then feel free to keep it with the updated generic device name!

dvrogozh replied:
I left this change in `.to()`.

albanD commented on torch/storage.py:
```diff
@@ -826,6 +859,13 @@ def hpu(self, device=None, non_blocking=False, **kwargs) -> T:  # type: ignore[m
         hpu_storage: torch.UntypedStorage = self._untyped_storage.hpu(device, non_blocking, **kwargs)
         return self._new_wrapped_storage(hpu_storage)
 
+    def to(self, device: torch.device, non_blocking: bool = False, **kwargs) -> T:  # type: ignore[type-var]
```
```suggestion
    def to(self, *, device: torch.device, non_blocking: bool = False) -> T:  # type: ignore[type-var]
```

I don't think we need the arbitrary kwargs here as this is a new API.
Using kwargs-only will allow us to add more arguments if needed in the future without breaking BC

dvrogozh replied:
done

albanD commented on torch/storage.py:
```diff
@@ -153,6 +186,9 @@ def _to(self, dtype):
             storage = storage.clone()
         return storage
 
+    def to(self, device: torch.device, non_blocking: bool = False, **kwargs) -> T:  # type: ignore[type-var] # noqa: E704
```
```suggestion
    def to(self, *, device: torch.device, non_blocking: bool = False) -> T:  # type: ignore[type-var] # noqa: E704
```
See below for details

dvrogozh replied:
done

albanD commented on torch/storage.py:
```diff
@@ -38,8 +38,41 @@ def size(self) -> int:
+                the ``non_blocking`` argument.
+        """
+        device2 = torch.device('cuda', device) if device else torch.device('cuda')
+        return self.to(device2, non_blocking, **kwargs)
```
The "async" argument here has been deprecated for a while, so I think we can take this apportunity to make this PR bc-breaking and remove `**kwargs` here altogether from the public APIs.

dvrogozh replied:
Yep. Same story - I worked on this PR in the assumption to preserve backward compatibility. If there is a decision to break it, I will make a change.

albanD replied:
Yes the "not changing it" is definitely the right default behavior for these PR.
This case is a bit different as it makes our new public API significantly worse. So I think it's worth going the slightly more risky route of changing the behavior.
Thanks for being careful in not breaking thing by the way, it is very appreciated.

dvrogozh replied:
Ok, appreciate if this can be done in a follow up PR. FYI, inside pytorch code only `validate_cuda_device` is called, here:
https://github.com/pytorch/pytorch/blob/3d20a5dd87e9c45e20dbc89210cfc63e7824b2b9/torch/jit/_serialization.py#L181-L182

I don't see calls to `validate_hpu_device`, though both are declared as public APIs and reported to documentation.

dvrogozh replied:
Ok, I dropped kwargs.

albanD commented on torch/serialization.py:
```diff
@@ -394,29 +329,32 @@ def _validate_privateuse1_device(location, backend_name):
+
+def validate_cuda_device(location):
+    return _validate_device(location, 'cuda').index
 
 
-def _privateuse1_deserialize(obj, location):
-    backend_name = torch._C._get_privateuse1_backend_name()
+def validate_hpu_device(location):
+    return _validate_device(location, 'hpu').index
```
Why can't these use the _validate_device function above? Is it to preserve public API?

cc @mikaylagawarecki should we deprecate/remove these public APIs?

dvrogozh replied:
> Why can't these use the _validate_device function above? Is it to preserve public API?

Yes, that's to preserve public API. Note that if `validate_cuda_device` and `validate_hpu_device` are removed, then BC-Linter notices this and fails. So, I decided to add these simple wrappers. If you believe it's time to break backward compatibility - let me know, I'll remove them. But you will need help me with silencing BC-Linter since I am not familiar with the process of how backward incompatible changes are managed in pytorch.

albanD replied:
We can do these in a follow up if @mikaylagawarecki wants it.
We can keep it as is for this PR for sure.

dvrogozh replied:
Ok, appreciate if this can be done in a follow up PR. FYI, inside pytorch code only `validate_cuda_device` is called, here:
https://github.com/pytorch/pytorch/blob/3d20a5dd87e9c45e20dbc89210cfc63e7824b2b9/torch/jit/_serialization.py#L181-L182

I don't see calls to `validate_hpu_device`, though both are declared as public APIs and reported to documentation.

albanD reviewed (commented):
Sounds good to me!

dvrogozh commented:
@albanD : as we anticipated, bc-linter fails on the last version after removing `kwargs` (https://github.com/pytorch/pytorch/actions/runs/9103745292/job/25032253108?pr=125530):
```
Warning: Function TypedStorage.cuda: **kwargs was removed
Warning: Function TypedStorage.hpu: **kwargs was removed
```

Please, let me know how this should be handled? Should `suppress-bc-linter` be added to PR? (note: I don't have such rights myself)

albanD reviewed (approved):
Thanks!

albanD commented:
@pytorchbot merge -r

dvrogozh commented:
@albanD : merge has failed due to some unrelated issues. Should anything be done manually to retrigger ci or bot will automatically pick it up again for merge later on?

albanD commented:
@pytorchbot merge -i

Nothing to do this was fixed in trunk.
